usecases/config/runtime/manager.go

```diff
@@ -0,0 +1,194 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package runtime
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	ErrEmptyConfig         = errors.New("empty runtime config")
+	ErrFailedToOpenConfig  = errors.New("failed to open runtime config")
+	ErrFailedToReadConfig  = errors.New("failed to read runtime config ")
+	ErrFailedToParseConfig = errors.New("failed to parse runtime config ")
+)
+
+// Parser takes care of unmarshaling a config struct
+// from given raw bytes(e.g: YAML, JSON, etc).
+type Parser[T any] func([]byte) (*T, error)
+
+// ConfigManager takes care of periodically loading the config from
+// given filepath for every interval period.
+type ConfigManager[T any] struct {
+	// path is file path of config to load and unmarshal from
+	path string
+	// interval is how often config manager trigger loading the config file.
+	interval time.Duration
+	// parse takes care of unmarshaling the config struct from a file
+	parse Parser[T]
+
+	// currentConfig is last successfully loaded config.
+	// ConfigManager keep using this config if there are any
+	// failures to load new configs.
+	currentConfig *T
+	currentHash   string
+	mu            sync.RWMutex // protects currentConfig
+
+	log             logrus.FieldLogger
+	lastLoadSuccess prometheus.Gauge
+	configHash      *prometheus.GaugeVec
+}
+
+func NewConfigManager[T any](
+	filepath string,
+	parser Parser[T],
+	interval time.Duration,
+	log logrus.FieldLogger,
+	r prometheus.Registerer,
+) (*ConfigManager[T], error) {
+	// catch empty filepath early
+	if len(strings.TrimSpace(filepath)) == 0 {
+		return nil, errors.New("filepath to load runtimeconfig is empty")
+	}
+
+	cm := &ConfigManager[T]{
+		path:     filepath,
+		parse:    parser,
+		interval: interval,
+		log:      log,
+		lastLoadSuccess: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_last_load_success",
+			Help: "Whether the last loading attempt of runtime config was success",
+		}),
```

Comment:
Instead of a boolean, we could store the last successful load timestamp into the gauge and then you can have alerts like to catch stale configurations:
```promql
(time() - weaviate_runtime_config_last_load_success) > 5 * 60 // it was not updated for 5 minutes
```

Reply:
good point. I have slightly different opinion here.

We can accomplish same alert with something like this.

```yaml
   alert: WeaviateBadRuntimeConfig
    expr: |
      # The metric value is reset to 0 on error while reloading the config at runtime.
      weaviate_runtime_config_last_reload_success == 0
    for: 5m
```

Plus having just boolean also doesn't require `time.Now()` syscall everytime.

I would have done the same thing like you mentioned if it's for proper cronJob or anything. But keeping it just boolean is simpler enough for this case I believe. wdyt?

Reply:
yea, makes sense. Just throwing ideas around